src/mechanics/mesh/MeshGmsh.cpp

```diff
@@ -0,0 +1,378 @@
+#include "MeshGmsh.h"
+
+#include "base/Exception.h"
+#include "mechanics/interpolation/InterpolationTrussLinear.h"
+#include "mechanics/interpolation/InterpolationTriangleLinear.h"
+#include "mechanics/interpolation/InterpolationQuadLinear.h"
+
+#include <array>
+#include <climits>
+#include <algorithm>
+#include <fstream>
+#include <functional>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <vector>
+
+
+// Helper structs (cpp only)
+// %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+struct GmshHeader
+{
+    double version;
+    bool isBinary;
+    int doubleSize;
+};
+
+struct GmshNode
+{
+    int id = 0;
+    std::array<double, 3> coordinates;
+};
+
+struct GmshElement
+{
+    int id = 0;
+    int type = 0;
+    std::vector<int> tags;
+    std::vector<int> nodes;
+};
+
+struct GmshPhysicalNames
+{
+    int id = 0;
+    int dimension = 0;
+    std::string name;
+};
+
+
+struct GmshFileContent
+{
+    GmshHeader header;
+    int minNodeId = INT_MAX;
+    int maxNodeId = INT_MIN;
+    int dimension = 1;
+    std::vector<GmshNode> nodes;
+    std::vector<GmshElement> elements;
+    std::vector<GmshPhysicalNames> physicalNames;
+};
+
+// Helper functions (cpp only)
+// %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+static GmshHeader ReadGmshHeader(std::ifstream& file)
+{
+    std::string line;
+    getline(file, line);
+
+    int binary;
+    GmshHeader header;
+    file >> header.version;
+    file >> binary;
+    file >> header.doubleSize;
+
+    header.isBinary = (binary == 1);
+
+    std::getline(file, line); // endl
+    if (header.version < 2.)
+    {
+        throw NuTo::Exception(__PRETTY_FUNCTION__, "Gmsh version 2.0 or higher requiered. - File version is " +
+                                                           std::to_string(header.version));
+    }
+    getline(file, line); // $EndMeshFormat
+    return header;
+}
+
+
+void ReadNodesASCII(std::istream& file, GmshFileContent& fileContent)
+{
+    std::string line;
+    std::getline(file, line);
+    int numNodes = std::stoi(line);
+
+    fileContent.nodes.resize(numNodes);
+    bool is2d = false;
+    bool is3d = false;
+    for (GmshNode& node : fileContent.nodes)
+    {
+        file >> node.id;
+        file >> node.coordinates[0];
+        file >> node.coordinates[1];
+        file >> node.coordinates[2];
+        fileContent.minNodeId = std::min(fileContent.minNodeId, node.id);
+        fileContent.maxNodeId = std::max(fileContent.maxNodeId, node.id);
+        is2d = (is2d || node.coordinates[1] != 0);
```

Comment:
Well yes, since you are reading directly from a file here. If there is a 0 in the file, than your double would also be 0. If the node in the file has a non zero value, even with minimal distance to 0, you probably messed something up in gmsh or shiftet your plane/line to this value (which is currently not allowed). ;)
We make an assumption/prescription about the meshes dimension here. 1d is always  [x,0,0] and 2d is always [x,y,0]. We did this to avoid a lot of checks. But that musn't be our final approach. What do you think, should we be more flexible for 1d and 2d mesh generation?

Reply:
Nope, seems fine as it is then.